recipe-client-addon/lib/PreferenceManagement.jsm

```diff
@@ -0,0 +1,395 @@
+   */
+  validateRecipePrefTypes(recipe) {
+    const { preferenceName, preferenceType } = recipe;
+
+    const prevPrefType = Services.prefs.getPrefType(preferenceName);
+    const givenPrefType = PREFERENCE_TYPE_MAP[preferenceType];
+
+    if (!preferenceType || !givenPrefType) {
+      throw new Error(`Invalid preferenceType provided (given "${preferenceType}")`);
+    }
+
+    if (prevPrefType !== Services.prefs.PREF_INVALID && prevPrefType !== givenPrefType) {
+      throw new Error(
+        `Previous preference value is of type "${prevPrefType}", but was given "${givenPrefType}" (${preferenceType})`
+      );
+    }
+  },
+
+  /**
+   * Saves a recipe object into locally-namespaced storage.
+   *
+   * @param  {Object} recipe    Recipe data to save into storage.
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async saveRecipeData(recipe) {
+    log.debug("PreferenceManagement.saveRecipeData()");
+
+    const store = await this.getStorage();
+    store.data[namespace][recipe.name] = recipe;
+    store.saveSoon();
+  },
+
+  /**
+   * Returns an array of stored recipe objects which are NOT expired.
+   *
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async getActiveChanges() {
+    log.debug("PreferenceManagement.getActiveChanges()");
+    const allPrefs = await this.getStoredRecipes();
+    // Return copies so mutating them doesn't affect the storage.
+    return allPrefs.filter(e => !e.expired).map(e => Object.assign({}, e));
+  },
+
+  /**
+   * Returns an array of stored recipe objects, both active and expired.
+   *
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async getStoredRecipes() {
+    log.debug("PreferenceManagement.getStoredRecipes()");
+    const store = await this.getStorage();
+
+    // Return copies so mutating them doesn't affect the storage.
+    return Object.values(store.data[namespace]).map(e => Object.assign({}, e));
+  },
+
+  /**
+   * Determines if the local storage has data for a specific recipe.
+   *
+   * @param  {string}  recipeName
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async has(recipeName) {
+    log.debug(`PreferenceManagement.has(${recipeName})`);
+    const store = await this.getStorage();
+    return store.data[namespace].hasOwnProperty(recipeName);
+  },
+
+  /**
+   * Expires a stored recipe.
+   *
+   * @param  {string}  recipeName
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async expire(recipeName) {
+    log.debug(`PreferenceManagement.expire(${recipeName})`);
+    const store = await this.getStorage();
+    store.data[namespace][recipeName].expired = true;
```

Comment:
Should this throw something more descriptive if the `[namespace][recipeName]` object doesn't exist?